crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+use std::io::{Cursor, Read};
+use std::sync::Arc;
+
+use itertools::Itertools;
+use zip::{read::ZipFile, result::ZipResult, ZipArchive};
+
+use crate::file_system::{FileRevision, FileSystem, FileSystemPath, FileType, Metadata, Result};
+
+#[derive(Debug, Clone)]
+pub struct VendoredFileSystem {
+    inner: VendoredZipArchive,
+}
+
+impl VendoredFileSystem {
+    const READ_ONLY_PERMISSIONS: u32 = 0o444;
+
+    pub fn new(raw_bytes: Arc<[u8]>) -> Result<Self> {
+        Ok(Self {
+            inner: VendoredZipArchive::new(raw_bytes)?,
+        })
+    }
+
+    /// Returns an owned copy of the underlying zip archive,
+    /// that can be queried for information regarding contained files and directories
+    fn vendored_archive(&self) -> VendoredZipArchive {
+        // Zip lookups require a mutable reference to the archive,
+        // which creates a nightmare with lifetimes unless you own the archive, hence the clone.
+        //
+        // The zip documentation states that:
+        //
+        //     At the moment, this type is cheap to clone if this is the case for the reader it uses.
+        //     However, this is not guaranteed by this crate and it may change in the future.
+        //
+        // So this should be fairly cheap *for now*.
+        self.inner.clone()
+    }
+
+    /// Create a new empty string, read the contents of `zipfile` into it, and return the string
+    fn read_zipfile(mut zipfile: ZipFile) -> Result<String> {
+        let mut buffer = String::with_capacity(zipfile.size() as usize);
+        zipfile.read_to_string(&mut buffer)?;
+        Ok(buffer)
```

Comment:
I think we can just use `read_to_string` here. I haven't followed all the code but `zipfile.read_to_string` calls into `std::read_to_string` and I'm pretty sure they correctly resize the buffer before writing into it.

Reply:
You have to pass in a pre-existing buffer for the `read_to_string` method

Reply:
Sorry. What I meant is we can simplify this to 

```rust
let mut buffer = String::new();
zipfile.read_to_string(&buffer)?;
Ok(buffer)
```

Because `read_to_string` already handles the `with_capacity` (it should call `reserve` or similar)

Reply:
Got it, thanks!